src/tir/schedule/primitive/compute_at.cc

```diff
@@ -0,0 +1,584 @@
+    const BlockNode* required_block = TVM_SREF_TO_BLOCK(required_block, required_block_sref);
+    ICHECK(block2realize.count(required_block));
+    RelaxBufferRegions</*relax_storage_scope=*/is_compute_at>(
+        /*binding=*/GetBindings(GetRef<BlockRealize>(block2realize.at(required_block))),
+        /*buffer_regions=*/is_compute_at ? required_block->reads : required_block->writes,
+        /*relax_path_low_inclusive=*/GetRef<StmtSRef>(required_block_sref->parent),
+        /*relax_path_high_exclusive=*/loop_sref, /*relaxed=*/required_regions);
+  }
+}
+
+/******** Main Implementation ********/
+
+template <bool is_compute_at>
+void ComputeAtOrReverseComputeAtImpl(ScheduleState self, const StmtSRef& block_sref,
+                                     const StmtSRef& loop_sref, bool preserve_unit_loops) {
+  const BlockNode* block = TVM_SREF_TO_BLOCK(block, block_sref);
+  const ForNode* loop = TVM_SREF_TO_FOR(loop, loop_sref);
+  // Step 1. Bunch of checks
+  // Check condition 1) and 2): stage pipeline and subtree compact dataflow
+  StmtSRef scope_root_sref = GetScopeRoot(self, block_sref,
+                                          /*require_stage_pipeline=*/true,
+                                          /*require_subtree_compact_dataflow=*/true);
+  Block scope_root = GetRef<Block>(scope_root_sref->StmtAs<BlockNode>());
+  BlockScope scope = self->GetBlockScope(scope_root_sref);
+  Array<StmtSRef> producer_srefs = GetProducers(block_sref, scope);
+  Array<StmtSRef> consumer_srefs = GetConsumers(block_sref, scope);
+  arith::Analyzer analyzer;
+  // Check condition 3): `block` and `loop` are under the same scope,
+  // and `loop` is not the ancestor of `block`
+  NotInSameScopeError::CheckAndBindLoopDomain(self, block_sref, loop_sref, scope_root_sref,
+                                              &analyzer);
+  // Check condition 4): `block` is not an output block
+  if (is_compute_at) {
+    CheckNotOutputBlock(self, block_sref, scope_root_sref);
+  }
+  // Step 2. Plan for the removal of `block`
+  ScopeReconstructor reconstructor(scope_root, GetRef<Block>(block), GetRef<For>(loop));
+  LeafBlockRemovalPlan(self, block_sref, &reconstructor.rm_src_stmt_, &reconstructor.rm_tgt_stmt_);
+  // Step 3. Find the insertion point under `loop`
+  // Check condition 5): all the required block are under the given loop
+  std::unordered_map<const BlockNode*, const BlockRealizeNode*> block2realize;
+  block2realize.reserve(self->block_info.size());
+  int insert_position = FindInsertionPoint<!is_compute_at, is_compute_at>(
+      /*self=*/self,
+      /*subtrees=*/AsArray(loop->body),
+      /*producer_srefs=*/producer_srefs,
+      /*consumer_srefs=*/consumer_srefs, /*block2realize=*/&block2realize);
+  // Step 4. Calculate the region provided by a single execution instance of `block`,
+  // as well as the region required by dependent blocks under `loop`
```

Comment:
I intentionally use the term `provide`/`require` instead of `produce`/`consume` to prevent you from such misunderstanding.

Here is the definition:
- In compute-at, `provide` means `produce` and `require` means `consume`
- In reverse-compute-at, `provide` means `consume` and `require` means `produce`

Reply:
probably I should make it crystal clear here in the doc

Reply:
Yeah I see the new document. Thanks!